clients/venice-thin-client/src/main/java/com/linkedin/venice/client/schema/RouterBackedSchemaReader.java

```diff
@@ -300,6 +298,309 @@ private SchemaEntry fetchSingleSchema(String requestPath, boolean isValueSchema)
     return new SchemaEntry(schemaResponse.getId(), writerSchema);
   }
 
+  /**
+   * This method will try to update all value schemas.
+   * Based on Router endpoint availability, it should first try the new method: (1) Fetch all value schema IDs first,
+   * (2) then fetch each individual schema one by one. If the forceRefresh is true, it will query all known value schema
+   * IDs, otherwise, it will only query value schemas that are not populated in cached schema map.
+   * If the new router endpoint /value_schema_ids/ is not available, it should fall back to query all value schemas at once.
+   */
+  private void updateAllValueSchemas(boolean forceRefresh) {
+    try {
+      Set<Integer> valueSchemaIdSet;
+      try {
+        valueSchemaIdSet = fetchAllValueSchemaIdsFromRouter();
+      } catch (Exception e) {
+        LOGGER.warn(
+            "Caught exception when trying to fetch all value schema IDs from router, will fetch all value schema entries instead.");
+        // Fall back to fetch all value schema.
+        for (SchemaEntry valueSchemaEntry: fetchAllValueSchemaEntriesFromRouter()) {
+          valueSchemaEntryMap.put(valueSchemaEntry.getId(), valueSchemaEntry);
+          valueSchemaMapR.put(valueSchemaEntry.getSchema(), valueSchemaEntry.getId());
+        }
+        return;
+      }
+
+      for (int id: valueSchemaIdSet) {
+        try {
+          maybeFetchValueSchemaEntryById(id, forceRefresh);
+        } catch (VeniceClientException e) {
+          LOGGER.error("Got exception while trying fetch schema id: " + id + " from router for store: " + storeName);
+        }
+      }
+
+    } catch (Exception ex) {
+      throw new VeniceClientException(
+          "Got exception while trying to fetch all value schemas for store: " + storeName,
```

Comment:
So the code would catch the exception for each individual schema fetch and the outer layer would try to re-throw the exception.
What is the expected behavior here? Is the schema refresh is a best effort or all or nothing?
I saw `maybeFetchValueSchemaEntryById` would update the current schema map directly, so it is not all or nothing , so maybe we can maintain the last exception and re-throw it after the loop?

Reply:
Umm, I think I get what you mean. The outer loop is older code in this PR, I can update this part.

Reply:
Oh because fetchAllValueSchemaEntriesFromRouter also need to be handled. But anyway let me try to adjust the handling in this method.